File: week-2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// function prototypes
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // get input text from the user
    string text = get_string("Text: ");

    // count letters, words, and sentences
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // calculate L (letters per 100 words) and S (sentences per 100 words)
    float L = (letters / (float) words) * 100;
    float S = (sentences / (float) words) * 100;

    // calculate coleman-liau index
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // print the grade level
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", (int) round(index));
    }

    return 0;
}

// count the number of letters in the text
int count_letters(string text)
{
    int count = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }

    return count;
}

// count the number of words in the text
int count_words(string text)
{
    int count = 1; // because words are separated by spaces

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]))
        {
            count++;
        }
    }

    return count;
}

// count the number of sentences in the text
int count_sentences(string text)
{
    int count = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }

    return count;
}

File: week-6/sentimental-readability/readability.py
def main():
    # get text from user
    text = input("Text: ")

    # count letters, words, and sentences
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # calculate coleman-liau index
    L = (letters / words) * 100
    S = (sentences / words) * 100
    index = round(0.0588 * L - 0.296 * S - 15.8)

    # print grade level
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


def count_letters(text):
    # count alphabetic characters in text
    return sum(1 for c in text if c.isalpha())


def count_words(text):
    # count spaces and add 1 (words = spaces + 1)
    return len(text.split())


def count_sentences(text):
    # count sentence endings (periods, exclamations, and question marks)
    return sum(1 for c in text if c in [".", "!", "?"])


if __name__ == "__main__":
    main()
